### PR TITLE
Add support for remote-store feature in OpenSearch

### PR DESCRIPTION
**Pull request by rishabh6788**
### Description
This PR adds support to enable remote-store feature in OpenSearch. See feature details [here](https://github.com/opensearch-project/OpenSearch/issues/1968)

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/8087. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by rishabh6788** [comment]
Currently blocked on https://github.com/opensearch-project/OpenSearch/issues/8158.

File: lib/infra/infra-stack.ts [hunk 1]
```diff
@@ -79,13 +83,20 @@ export class InfraStack extends Stack {
...
         ManagedPolicy.fromAwsManagedPolicyName('CloudWatchAgentServerPolicy'),
         ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore')],
       assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
     });
 
+    if (props.enableRemoteStore) {
+      // Remote Store needs an S3 bucket to be registered as snapshot repo
+      // Add scoped bucket policy to the instance role attached to the EC2
+      const remoteStoreObj = new RemoteStoreResources(this);
+      this.instanceRole.addToPolicy(remoteStoreObj.getRemoteStoreBucketPolicy());
```
Comment (peterzhuamazon) [inline lib/infra/infra-stack.ts:97]: Is this bucket fixed in cdk cluster specifically?

File: lib/infra/infra-stack.ts [hunk 2]
```diff
@@ -457,6 +468,28 @@ export class InfraStack extends Stack {
           ignoreErrors: false,
         }));
       }
+
+      if (props.enableRemoteStore) {
+        if (props.distributionUrl.includes('artifacts.opensearch.org') && !props.minDistribution) {
+          cfnInitConfig.push(InitCommand.shellCommand('set -ex;cd opensearch; echo "y"|sudo -u ec2-user bin/opensearch-plugin install repository-s3', {
```
Comment (peterzhuamazon) [inline lib/infra/infra-stack.ts:474]: So can we use `--batch` instead of `echo yes`?

File: lib/infra/infra-stack.ts [hunk 3]
```diff
@@ -457,6 +468,28 @@ export class InfraStack extends Stack {
...
+            cwd: '/home/ec2-user',
+            ignoreErrors: false,
+          }));
+        } else {
+          cfnInitConfig.push(InitCommand.shellCommand('set -ex;cd opensearch; echo "y"|sudo -u ec2-user bin/opensearch-plugin install '
```
Comment (peterzhuamazon) [inline lib/infra/infra-stack.ts:479]: Same as above.

File: lib/infra/remote-store-resources.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    constructor(scope: Stack) {
+      this.snapshotS3Bucket = new Bucket(scope, `remote-store-${scope.stackName}`, {
+        removalPolicy: RemovalPolicy.DESTROY,
+        autoDeleteObjects: true,
+        bucketName: `${scope.stackName}`,
+      });
+
+      this.bucketPolicyStatement = new PolicyStatement({
+        effect: Effect.ALLOW,
+        actions: ['s3:*'],
```
Comment (peterzhuamazon) [inline lib/infra/remote-store-resources.ts:19]: Is this too open for the bucket?

**Review by peterzhuamazon — COMMENTED** [review]
Hi @rishabh6788 add some comments regarding changes.
Thanks.

File: lib/infra/infra-stack.ts (hunk 1 above)
Comment (rishabh6788) [reply lib/infra/infra-stack.ts:97]: Yes, the bucket is created during cdk deployment and deleted when the stack is deleted or destroyed.

File: lib/infra/infra-stack.ts (hunk 2 above)
Comment (rishabh6788) [reply lib/infra/infra-stack.ts]: Wasn't aware of this, updated. 
Thanks!

File: lib/infra/infra-stack.ts (hunk 3 above)
Comment (rishabh6788) [reply lib/infra/infra-stack.ts]: Updated.

File: lib/infra/remote-store-resources.ts (hunk 4 above)
Comment (rishabh6788) [reply lib/infra/remote-store-resources.ts]: Makes sense, updated bucket policy as per `repository-s3` plugin recommendation.

**Pull request merged**